c2rust-refactor/src/scripting/to_lua_ast_node.rs

```diff
@@ -286,15 +375,211 @@ unsafe impl Send for LuaAstNode<P<Expr>> {}
+                ExprKind::Assign(lhs, rhs) => Ok(vec![LuaAstNode::new(lhs.clone()), LuaAstNode::new(rhs.clone())]),
+                ExprKind::Call(func, params) => {
+                    let mut exprs = Vec::with_capacity(params.len() + 1);
+
+                    exprs.push(LuaAstNode::new(func.clone()));
+
+                    for param in params {
+                        exprs.push(LuaAstNode::new(param.clone()));
+                    }
+
+                    Ok(exprs)
+                },
+                _ => Ok(Vec::new()),
+            }
+        });
+
+        methods.add_method("set_exprs", |_lua_ctx, this, exprs: Vec<LuaAstNode<P<Expr>>>| {
+            match &mut this.borrow_mut().node {
+                ExprKind::Field(expr, _)
+                | ExprKind::Unary(_, expr) => *expr = exprs[0].borrow().clone(),
+                ExprKind::MethodCall(_, exprs) => *exprs = exprs.iter().map(|e| e.clone()).collect(),
+                ExprKind::Call(func, params) => {
+                    *func = exprs[0].borrow().clone();
+                    *params = exprs.iter().skip(1).map(|e| e.borrow().clone()).collect()
+                },
+                ExprKind::Assign(lhs, rhs)
+                | ExprKind::Binary(_, lhs, rhs) => {
+                    *lhs = exprs[0].borrow().clone();
+                    *rhs = exprs[1].borrow().clone();
+                },
+                ExprKind::Cast(expr, _) => *expr = exprs[0].borrow().clone(),
+                e => unimplemented!("LuaAstNode<P<Expr>>:set_exprs() for {}", e.ast_name()),
+            }
+
+            Ok(())
+        });
+
+        methods.add_method("get_op", |_lua_ctx, this, ()| {
+            match &this.borrow().node {
+                ExprKind::Unary(op, _) => Ok(Some(op.ast_name())),
+                // ExprKind::Binary(op, ..) |
+                // ExprKind::AssignOp(op, ..) => Ok(op.ast_name()),
+                _ => Ok(None),
+            }
+        });
+
+        methods.add_method("print", |_lua_ctx, this, ()| {
+            println!("{:?}", this.borrow());
+
+            Ok(())
+        });
+
+        methods.add_method("get_method_name", |lua_ctx, this, ()| {
+            if let ExprKind::MethodCall(path_seg, ..) = &this.borrow().node {
+                return path_seg.ident.as_str().get().to_lua(lua_ctx).map(|s| Some(s))
+            }
+
+            Ok(None)
+        });
+
+        methods.add_method("set_method_name", |_lua_ctx, this, name: LuaString| {
+            if let ExprKind::MethodCall(path_seg, ..) = &mut this.borrow_mut().node {
+                path_seg.ident = Ident::from_str(name.to_str()?);
+            }
+
+            Ok(())
+        });
+
+        methods.add_method("get_lit", |_lua_ctx, this, ()| {
+            if let ExprKind::Lit(lit) = &this.borrow().node {
+                return Ok(Some(LuaAstNode::new(lit.clone())))
+            }
+
+            Ok(None)
+        });
+
+        methods.add_method("to_lit", |_lua_ctx, this, lit: LuaAstNode<Lit>| {
```

Comment:
I’m not sure about all these `to_*` methods. Why not use a single `set` method that just replaces the node?

Reply:
How would the set method know what what kind of node you're creating?

IE for lit you need a literal, for binary you need an op and two exprs. A generic set method would, I think, only work if it takes an expr and replaces self with it? But then you'd still need to create that expr somehow?

Reply:
Ok, I get what you're saying. Basically these are just `ExprKind` constructors. I guess it's fine for them to be on `Expr` then.